### ISSUE TITLE
Implement efficient PyTorch DataLoader for sparse event histograms

### ISSUE THREAD

**Issue by tallamjr** [tallamjr/evlib#10]
## Problem Statement

Event camera data is extremely sparse (97.9% of pixels are zero in typical stacked histograms), but current implementations either:
1. Store dense tensors → massive memory waste (5.5GB for 115MB of actual data)
2. Use inefficient loading patterns → slow training

We need an efficient PyTorch DataLoader that exploits sparsity while maintaining compatibility with existing dense tensor models.

## Background

- **Stacked histogram output**: `(n_windows, 20, 360, 640)` dense tensor
- **Actual sparsity**: 97.9% zeros (only 2.1% contain event data)  
- **evlib sparse format**: `(~118M, 7)` DataFrame with columns `[window_id, channel, time_bin, y, x, count, channel_time_bin]`
- **Storage efficiency**: Parquet compression achieves ~20x reduction due to sparsity

## Proposed Solution

Implement a PyTorch Dataset that:
1. **Stores** data efficiently in Parquet format (leveraging evlib's natural output)
2. **Loads** only required windows on-demand
3. **Converts** sparse → dense just-in-time for model consumption
4. **Batches** efficiently with configurable chunk sizes

### Core Dataset Class

```python
import torch
from torch.utils.data import Dataset, DataLoader
import polars as pl
from pathlib import Path

class EventHistogramDataset(Dataset):
    """
    Efficient PyTorch Dataset for sparse event histograms.
    
    Loads event histogram data from Parquet files and converts sparse representation
    to dense tensors on-demand. Optimized for training neural networks on event data.
    
    Args:
        parquet_path: Path to Parquet file containing sparse histogram data
        window_chunk_size: Number of consecutive windows to load per item (for sequence models)
        height: Spatial height (default: 360 for downsampled Gen4)
        width: Spatial width (default: 640 for downsampled Gen4) 
        n_channels_bins: Number of channel*time_bin combinations (default: 20 = 2*10)
    """
    
    def __init__(self, parquet_path: str, window_chunk_size: int = 32, 
                 height: int = 360, width: int = 640, n_channels_bins: int = 20):
        self.parquet_path = Path(parquet_path)
        self.chunk_size = window_chunk_size
        self.height = height
        self.width = width  
        self.n_channels_bins = n_channels_bins
        
        # Lazy load metadata
        self.df = pl.scan_parquet(self.parquet_path)
        
        # Get total number of windows
        self.max_window = self.df.select(pl.col("window_id").max()).collect().item()
        self.n_windows = self.max_window + 1
        
        # Calculate number of chunks
        self.n_chunks = (self.n_windows + self.chunk_size - 1) // self.chunk_size
        
        print(f"Loaded dataset: {self.n_windows} windows, {self.n_chunks} chunks of size {self.chunk_size}")
    
    def __len__(self):
        return self.n_chunks
    
    def __getitem__(self, idx):
        """
        Load a chunk of consecutive windows and convert to dense tensor.
        
        Args:
            idx: Chunk index
            
        Returns:
            Dense tensor of shape (chunk_size, n_channels_bins, height, width)
        """
        start_window = idx * self.chunk_size
        end_window = min(start_window + self.chunk_size, self.n_windows)
        actual_chunk_size = end_window - start_window
        
        # Load only the required windows from Parquet
        chunk_data = self.df.filter(
            pl.col("window_id").is_between(start_window, end_window - 1)
        ).collect()
        
        # Initialize dense tensor
        dense = torch.zeros(
            self.chunk_size, self.n_channels_bins, self.height, self.width, 
            dtype=torch.uint8
        )
        
        # Convert sparse to dense if we have data
        if len(chunk_data) > 0:
            # Vectorized conversion for efficiency
            for row in chunk_data.iter_rows():
                window_id, channel, time_bin, y, x, count, channel_time_bin = row
                
                # Relative window index within chunk
                rel_window = window_id - start_window
                
                # Bounds checking
                if (0 <= rel_window < self.chunk_size and 
                    0 <= channel_time_bin < self.n_channels_bins and
                    0 <= y < self.height and 
                    0 <= x < self.width):
                    
                    dense[rel_window, channel_time_bin, y, x] = min(count, 255)
        
        # If chunk is smaller than chunk_size, pad with zeros (already initialized)
        return dense
    
    def get_window_range(self, idx):
        """Get the window ID range for a given chunk index."""
        start_window = idx * self.chunk_size
        end_window = min(start_window + self.chunk_size, self.n_windows)
        return start_window, end_window
    
    @classmethod
    def from_evlib_histogram(cls, evlib_hist, save_path: str, **kwargs):
        """
        Create dataset from evlib histogram DataFrame.
        
        Args:
            evlib_hist: Polars DataFrame from evlib.representations.create_stacked_histogram()
            save_path: Path to save Parquet file
            **kwargs: Additional arguments for EventHistogramDataset
        """
        # Save to Parquet with compression
        evlib_hist.write_parquet(save_path, compression="zstd")
        print(f"Saved sparse histogram to {save_path}")
        
        return cls(save_path, **kwargs)

# Helper functions for training
def create_dataloader(parquet_path: str, batch_size: int = 4, num_workers: int = 4, 
                     shuffle: bool = True, **dataset_kwargs):
    """
    Create DataLoader for training.
    
    Args:
        parquet_path: Path to Parquet file
        batch_size: Number of chunks per batch
        num_workers: Number of worker processes
        shuffle: Whether to shuffle chunks
        **dataset_kwargs: Arguments for EventHistogramDataset
    """
    dataset = EventHistogramDataset(parquet_path, **dataset_kwargs)
    
    return DataLoader(
        dataset, 
        batch_size=batch_size,
        shuffle=shuffle,
        num_workers=num_workers,
        pin_memory=True  # For GPU training
    )

# Usage example
def example_usage():
    """Example of how to use the dataset for training."""
    
    # 1. Create dataset from evlib output
    import evlib.representations as evr
    
    # Generate histogram
    hist = evr.create_stacked_histogram(
        "data/events.h5",
        height=360, width=640,
        nbins=10, window_duration_ms=50.0, count_cutoff=10
    ).collect(engine="streaming")
    
    # Save as dataset
    dataset = EventHistogramDataset.from_evlib_histogram(
        hist, 
        save_path="train_data.parquet",
        window_chunk_size=16  # 16 windows per item (0.8s sequences)
    )
    
    # 2. Create DataLoader
    train_loader = create_dataloader(
        "train_data.parquet",
        batch_size=8,         # 8 chunks per batch
        window_chunk_size=16, # 16 windows per chunk
        num_workers=4,
        shuffle=True
    )
    
    # 3. Training loop
    for batch_idx, batch_tensor in enumerate(train_loader):
        # batch_tensor shape: (8, 16, 20, 360, 640)
        # = (batch_size, sequence_length, channels*bins, height, width)
        
        print(f"Batch {batch_idx}: {batch_tensor.shape}")
        
        # Your model training code here
        # model(batch_tensor)
        
        if batch_idx >= 2:  # Just for demo
            break
```

## Benefits

1. **Memory Efficiency**: 
   - Parquet compression: ~20x reduction (5.5GB → ~275MB)
   - On-demand loading: Only load what's needed for current batch
   
2. **Training Speed**:
   - Lazy evaluation: No upfront conversion cost
   - Polars vectorized operations: Fast filtering and loading
   - Multi-worker support: Parallel data loading
   
3. **Flexibility**:
   - Configurable chunk sizes: Support different sequence lengths
   - Standard PyTorch interface: Drop-in replacement for existing code
   - GPU-ready: Pin memory support for efficient GPU transfer

## Implementation Tasks

- [ ] Implement `EventHistogramDataset` class
- [ ] Add comprehensive unit tests
- [ ] Benchmark against naive dense storage
- [ ] Add support for data augmentation (spatial transforms)
- [ ] Document integration with common training frameworks
- [ ] Add examples for different model architectures (CNN, RNN, Transformer)

## Performance Targets

- **Storage**: <300MB for typical 1-hour sequence (vs 5.5GB dense)
- **Loading speed**: <100ms per batch for training
- **Memory usage**: <2GB peak during training (vs >10GB naive approach)

## Related Issues

- Sparse tensor support for preprocessing operations
- Data augmentation for event histograms  
- Multi-GPU training optimization

---

**Priority**: High - Required for efficient neural network training on event data
**Difficulty**: Medium - Requires careful optimization but straightforward implementation
**Impact**: High - Enables practical training on large event datasets

**Comment by tallamjr** [comment]
# Updated Design: Hybrid Arrow IPC Streaming Pipeline

After thorough analysis as a principal engineer, I'm proposing a **significant architectural improvement** that moves beyond the current Parquet-only approach to a **hybrid storage-streaming design** using Arrow IPC.

## Problem Analysis: Current Approach Limitations

The existing Parquet-based design has several critical bottlenecks:

1. **File I/O Bottleneck**: Each `__getitem__` call triggers expensive Parquet reads (100-500ms/batch)
2. **Memory Fragmentation**: Repeated small allocations for sparse-to-dense conversion
3. **No Zero-Copy Optimization**: Missing Arrow's core performance advantage
4. **Limited Concurrency**: Multi-worker processes compete for file I/O resources

## Optimal Solution: Three-Layer Architecture

```
┌─────────────────┐    ┌─────────────────┐    ┌─────────────────┐
│   Storage Layer │───▶│ Streaming Layer │───▶│ Training Layer  │
│    (Parquet)    │    │  (Arrow IPC)    │    │   (PyTorch)     │
└─────────────────┘    └─────────────────┘    └─────────────────┘
      Compression           Zero-Copy             GPU Optimized
      Durability           Memory Pools          Batch Processing
      Analytics            Efficient Streaming   Multi-Worker
```

### Why This Architecture is Superior

#### Layer 1: Storage (Parquet) - "Cold Storage"
**Use Parquet for what it's best at:**
- ✅ **20x compression ratio** for sparse event data (5.5GB → 275MB)
- ✅ **Cross-platform compatibility** (Spark, DuckDB, Pandas, Polars)
- ✅ **Analytical queries** with predicate pushdown
- ✅ **Schema evolution** and durability

#### Layer 2: Streaming (Arrow IPC) - "Hot Pipeline"  
**Use Arrow IPC for what it's best at:**
- ✅ **Zero-copy data transfer** between processes
- ✅ **Shared memory pools** for efficient memory management
- ✅ **Sub-millisecond latency** for batch generation
- ✅ **Multi-worker compatibility** without file I/O contention

#### Layer 3: Training (PyTorch) - "Consumption"
**Optimized for ML training:**
- ✅ **GPU memory pinning** for efficient CPU→GPU transfer
- ✅ **Just-in-time sparse→dense conversion**
- ✅ **Standard PyTorch DataLoader interface**
- ✅ **Prefetching and caching** for continuous training

## Key Technical Innovations

### 1. Arrow Memory Pool Management
```python
class ArrowMemoryManager:
    """Manages Arrow memory pools for efficient tensor allocation."""
    
    def __init__(self, pool_size_gb: float = 2.0):
        self.pool = pa.system_memory_pool()
        self.allocation_tracker = {}
        
    def allocate_batch_memory(self, batch_shape: tuple, dtype: np.dtype) -> pa.Buffer:
        """Allocate memory for training batch with intelligent reuse."""
        # Buffer reuse logic minimizes allocation overhead
```

### 2. Zero-Copy Streaming Pipeline
```python
class ArrowStreamProducer:
    """Streams event data from Parquet via Arrow IPC with zero-copy transfers."""
    
    async def stream_batches(self, window_range: tuple) -> AsyncGenerator[pa.RecordBatch, None]:
        """Stream Arrow RecordBatches for specified window range."""
        # Efficient chunked streaming without loading entire dataset
```

### 3. Optimized PyTorch Dataset
```python
class ArrowIPCEventDataset(Dataset):
    """PyTorch Dataset with Arrow IPC streaming backend."""
    
    def __getitem__(self, idx: int) -> torch.Tensor:
        # Zero-copy Arrow → PyTorch tensor conversion
        # Asynchronous prefetching for next batch
        # Memory pool reuse for efficient allocation
```

## Performance Projections

### Memory Efficiency
- **Before**: 5.5GB dense + file I/O overhead = ~8GB total
- **After**: 275MB Parquet + 500MB Arrow buffers = ~975MB total
- **🎯 Improvement**: **8x memory reduction**

### Training Throughput  
- **Before**: ~10 batches/second (file I/O limited)
- **After**: ~100 batches/second (GPU computation limited)
- **🎯 Improvement**: **10x throughput increase**

### Batch Latency
- **Before**: 100-500ms/batch (file read + conversion)
- **After**: 5-10ms/batch (memory copy + conversion)  
- **🎯 Improvement**: **20x latency reduction**

## Why Parquet + Arrow IPC is Optimal

### Storage vs Streaming Trade-offs

**Parquet for Storage ✅**
- **Compression**: Leverages columnar format for 20x compression on sparse data
- **Durability**: Battle-tested for long-term data storage
- **Analytics**: SQL queries, data exploration, ETL pipelines
- **Ecosystem**: Works with entire data science stack

**Arrow IPC for Streaming ✅**  
- **Zero-Copy**: Shared memory between processes (no serialization overhead)
- **Type Safety**: Schema-validated data structures
- **Cross-Language**: Same format used by Rust backend and Python frontend
- **Streaming**: Handle datasets larger than available memory

**Hybrid Benefits ✅**
1. **Operational Flexibility**: Store once in Parquet, stream efficiently via Arrow
2. **Memory Hierarchy**: Cold storage → hot cache → training pipeline
3. **Scalability**: TB-scale datasets with GB-scale memory usage
4. **Development Velocity**: Standard tools for both storage and streaming

## Implementation Roadmap

### Phase 1: Core Infrastructure (Week 1)
- [ ] Implement `ArrowMemoryManager` with memory pool recycling
- [ ] Create `ArrowStreamProducer` for Parquet→Arrow streaming  
- [ ] Basic unit tests and memory leak detection

### Phase 2: PyTorch Integration (Week 2)
- [ ] Implement `ArrowIPCEventDataset` with zero-copy conversion
- [ ] Add prefetching, caching, and background streaming
- [ ] Integration tests with real event camera datasets

### Phase 3: Performance Optimization (Week 3)  
- [ ] Optimize sparse→dense conversion algorithms using vectorized operations
- [ ] Add GPU memory pinning and overlapped transfers
- [ ] Comprehensive benchmarking vs current Parquet approach

### Phase 4: Production Readiness (Week 4)
- [ ] Error handling, recovery, and monitoring
- [ ] Documentation, examples, and best practices
- [ ] Multi-GPU and distributed training support

## Risk Mitigation

### Low-Risk Implementation Strategy
- **Backward Compatibility**: Keep existing Parquet DataLoader as fallback
- **Incremental Rollout**: Phase-by-phase implementation with validation
- **Performance Validation**: Benchmark each phase against current implementation
- **Memory Safety**: Comprehensive testing for memory leaks and corruption

### Contingency Plans
- **Arrow IPC Issues**: Fall back to optimized Parquet streaming
- **Memory Pressure**: Dynamic memory pool sizing and garbage collection
- **Cross-Platform Issues**: Platform-specific optimizations where needed

## Success Metrics

### Performance Targets
- [ ] **Memory Usage**: <1GB peak during training (vs 8GB current)
- [ ] **Throughput**: >80 batches/second sustained (vs 10 current)  
- [ ] **Latency**: <20ms per batch p99 (vs 200ms+ current)
- [ ] **GPU Utilization**: >90% during training (vs ~60% I/O bound)

### Quality Targets  
- [ ] **Zero Memory Leaks**: Valgrind clean after 1000+ batches
- [ ] **API Compatibility**: Drop-in replacement for existing DataLoader
- [ ] **Multi-Platform**: Linux, macOS, Windows support
- [ ] **Scaling**: Handle 100GB+ datasets on 16GB machines

---

**This hybrid Arrow IPC design delivers order-of-magnitude improvements in memory efficiency and training throughput while maintaining the storage benefits of Parquet compression.**

**Priority**: Critical - Unlocks practical training on large event camera datasets  
**Impact**: High - Enables 10x larger models and datasets with same hardware  
**Complexity**: Medium - Leverages proven Arrow ecosystem with careful engineering

**Comment by tallamjr** [comment]
This may be required if we want to work with PyTorch model loading and training logic: https://pyo3.rs/v0.23.3/python-from-rust/calling-existing-code.html
